clang/lib/Driver/Driver.cpp

```diff
@@ -3306,10 +3282,13 @@ void Driver::handleArguments(Compilation &C, DerivedArgList &Args,
       if (InputArg->isClaimed())
         continue;
 
-      // Fortran input is preprocessed using the frontend.
-      if (InitialPhase == phases::FortranFrontend &&
+#ifdef ENABLE_CLASSIC_FLANG
+      // If the input is detected as already preprocessed (e.g. has the .f95
+      // extension), and the user specifies -E, preprocess the file anyway.
+      if (IsFlangMode() && InitialPhase == phases::Compile &&
```

Comment:
I'm not a Fortran expert, but `gfortran` docs suggest that 
* `.f95` means that the file _has not been preprocessed yet_

See here: https://gcc.gnu.org/onlinedocs/gcc/Overall-Options.html

Also, IIUC, this block of code makes sure that the parsing of preprocessor specific options is not happening here, i.e. in the compiler driver. Instead, it should happen in the frontend driver, i.e. where the preprocessor is run. So it's not about whether the preprocessor is run or not, but about whether to further parse the options or not.

I might be wrong though!

Reply:
@banach-space In the page you linked, the description for `.f95` says:

> <dl>
> <dt>*file*.f95</dt>
> <dd>Free form Fortran source code that should not be preprocessed.</dd>
> </dl>

The description is consistent with original Classic Flang code, as well as this patch. The uppercase `.F95` indicates that a file must be preprocessed, and the output of the preprocessor is typically named with `.f95`.

This block of code is a hack to maintain backward compatibility with legacy behaviour, i.e. allow `-E` to force preprocessing on files that are normally not preprocessed. If we can reach a consensus about not supporting this behaviour, I would prefer to get rid of this hack, and let the driver print a "previously preprocessed input" warning and do nothing. However, that is a bigger decision and should not be done in this PR.

Reply:
I misread that, sorry! This clarifies a lot @bryanpkc, thank you !

This makes sense - we should preserve the current behavior.